File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/preparation/ExoPlayerPlayableFilePreparationSourceProvider.java
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation;

import android.content.Context;
import android.os.Handler;
import com.google.android.exoplayer2.DefaultLoadControl;
import com.google.android.exoplayer2.LoadControl;
import com.google.android.exoplayer2.RenderersFactory;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelector;
import com.lasthopesoftware.bluewater.client.library.items.media.files.uri.BestMatchUriProvider;
import com.lasthopesoftware.bluewater.client.playback.engine.exoplayer.AudioRenderersFactory;
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.IPlayableFilePreparationSourceProvider;
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.mediasource.SpawnMediaSources;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PlayableFilePreparationSource;
import com.namehillsoftware.lazyj.CreateAndHold;
import com.namehillsoftware.lazyj.Lazy;
import org.joda.time.Minutes;


public class ExoPlayerPlayableFilePreparationSourceProvider implements IPlayableFilePreparationSourceProvider {

	private static final CreateAndHold<Integer> maxBufferMs = new Lazy<>(() -> (int) Minutes.minutes(5).toStandardDuration().getMillis());
	private static final CreateAndHold<TrackSelector> trackSelector = new Lazy<>(ExoPlayerPlayableFilePreparationSourceProvider::getNewTrackSelector);
	private static final CreateAndHold<LoadControl> loadControl = new Lazy<>(ExoPlayerPlayableFilePreparationSourceProvider::getNewLoadControl);

	private final Handler handler;
	private final BestMatchUriProvider bestMatchUriProvider;
	private final SpawnMediaSources mediaSourceProvider;
	private final RenderersFactory renderersFactory;

	public ExoPlayerPlayableFilePreparationSourceProvider(Context context, Handler handler, SpawnMediaSources mediaSourceProvider, BestMatchUriProvider bestMatchUriProvider) {
		this.handler = handler;
		this.bestMatchUriProvider = bestMatchUriProvider;

		this.mediaSourceProvider = mediaSourceProvider;

		renderersFactory = new AudioRenderersFactory(context);
	}

	@Override
	public int getMaxQueueSize() {
		return 1;
	}

	@Override
	public PlayableFilePreparationSource providePlayableFilePreparationSource() {
		return new ExoPlayerPlaybackPreparer(
			mediaSourceProvider,
			trackSelector.getObject(),
			loadControl.getObject(),
			renderersFactory,
			handler,
			bestMatchUriProvider);
	}

	private static TrackSelector getNewTrackSelector() {
		return new DefaultTrackSelector();
	}

	private static LoadControl getNewLoadControl() {
		final DefaultLoadControl.Builder builder = new DefaultLoadControl.Builder();
		builder.setBufferDurationsMs(
			DefaultLoadControl.DEFAULT_MIN_BUFFER_MS,
			maxBufferMs.getObject(),
			DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_MS,
			DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS);

		return builder.createDefaultLoadControl();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/preparation/ExoPlayerPlayableFilePreparationSourceProvider.kt
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation

import android.content.Context
import android.os.Handler
import androidx.media3.common.util.UnstableApi
import androidx.media3.exoplayer.DefaultLoadControl
import com.lasthopesoftware.bluewater.client.browsing.files.uri.BestMatchUriProvider
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.IPlayableFilePreparationSourceProvider
import com.lasthopesoftware.bluewater.client.playback.exoplayer.ExoPlayerProvider
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.buffering.BufferingExoPlayerProvider
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.mediasource.SpawnMediaSources
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.rendering.AudioRenderersFactory
import com.lasthopesoftware.resources.executors.HandlerExecutor
import org.joda.time.Minutes

@UnstableApi class ExoPlayerPlayableFilePreparationSourceProvider(
	private val context: Context,
	private val playbackHandler: Handler,
	private val interactionsHandler: HandlerExecutor,
	private val mediaSourceProvider: SpawnMediaSources,
	private val bestMatchUriProvider: BestMatchUriProvider
) : IPlayableFilePreparationSourceProvider {
	companion object {
		private val maxBufferMs by lazy { Minutes.minutes(5).toStandardDuration().millis.toInt() }
	}

	private val renderersFactory = AudioRenderersFactory(context)

	private val loadControl by lazy {
		val builder = DefaultLoadControl.Builder()
		builder
			.setBufferDurationsMs(
				DefaultLoadControl.DEFAULT_MIN_BUFFER_MS,
				maxBufferMs,
				DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_MS,
				DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS)
			.setPrioritizeTimeOverSizeThresholds(true)
		builder.build()
	}

	private val exoPlayerProvider by lazy {
		ExoPlayerProvider(
			context,
			renderersFactory,
			loadControl,
			interactionsHandler,
			playbackHandler
		)
	}

	private val bufferingExoPlayerProvider by lazy {
		BufferingExoPlayerProvider(interactionsHandler)
	}

    override val maxQueueSize get() = 1

	override fun providePlayableFilePreparationSource() = ExoPlayerPlaybackPreparer(
		mediaSourceProvider,
		exoPlayerProvider,
		bufferingExoPlayerProvider,
		bestMatchUriProvider
	)
}
